runtime/compiler/p/codegen/PPCPrivateLinkage.cpp

```diff
@@ -1913,6 +1913,27 @@ int32_t TR::PPCPrivateLinkage::buildPrivateLinkageArgs(TR::Node
 	    addDependency(dependencies, NULL, (TR::RealRegister::RegNum)i, TR_VSX_SCALAR, cg());
       }
 
+   // All CHelper Direct Calls on platforms other than Linux 32bit must load the target Address, in this case the helper's address in r12 (Linux 64LE) or 
+   // load the address of the jitTOC into r2 (Linux 64BE, and AIX)
+   if(linkage == TR_CHelper)
+   {
+      if(TR::Compiler->target.isLinux() && TR::Compiler->target.is64Bit() && TR::Compiler->target.cpu.isLittleEndian())
+      {
+         uintptrj_t targetAddress = (uintptrj_t)callSymbol->getMethodAddress();
+         loadConstant(cg(), callNode, (int64_t)targetAddress,dependencies->searchPreConditionRegister(TR::RealRegister::gr12));
+      }
+      else if ((TR::Compiler->target.isAIX() || TR::Compiler->target.isLinux()) && TR::Compiler->target.is64Bit())
+      {
+         generateTrg1MemInstruction(cg(),TR::InstOpCode::Op_load, callNode, dependencies->searchPreConditionRegister(TR::RealRegister::gr2), 
+            new(trHeapMemory()) TR::MemoryReference(cg()->machine()->getPPCRealRegister(TR::RealRegister::gr15), offsetof(J9VMThread, jitTOC), TR::Compiler->om.sizeofReferenceAddress(), cg()));
+      }
+      else if (TR::Compiler->target.isAIX() && TR::Compiler->target.is32Bit())
```

Comment:
We can also have Linux BE 32-bit (just double checked with @zl-wang). It also has J9VMThread in gpr13.  Please put this summary into the design and change the code accordingly:

**AIX 32-bit BE**: set gpr2 with TOC from  VMthread in gpr13
**AIX 64-bit BE**: set gpr2 with TOC from  VMThread in gpr15
**Linux 32-bit BE**: set gpr2 with TOC from  VMThread in gpr13
**Linux 64-bit LE**: set  gpr12 with target address
**Linux 64-bit BE**: set gpr2 with TOC from  VMThread in gpr15

I hope the table is correct, but @zl-wang please double check.

Reply:
Sorry, I only added a confusion. Linux 32-bit BE does not have TOC, so we should not set gpr2. @zl-wang. I think having a comment in the code and a summary in the description would've helped.

Reply:
The J9VMThread register should be retrieved by cg->getMethodMetaDataRegister().   It is gr13 on 32bit JVM and gr15 on 64bit JVM respectively.  It is a locked down register (dedicated register). It can be in dependency condition, just not depending on any other real register.

There are many examples of using method metadata register.

Reply:
AIX 32-bit BE:
AIX 64-bit BE:
Linux 64-bit BE:
       These are aix-style linkage.  There is a TOC-base register (gr2) from function descriptor if that function is called.  You can load it from jitTOC field of J9VMThread (pointed to by metaDataRegister).

Linux 32bit BE:   no TOC.  Don't touch gr2.  gr2 is reserved.

Linux 64-bit LE:   there is TOC (gr2), but there is no function descriptor.  TOC-base set-up is defined by ABI to go through a function's global-entry (relocation).  The global-entry address should be set up in gr12.